### PR TITLE
docs/hinted_handoff: Describe sync point HTTP API

### PR DESCRIPTION
**Pull request by dawmd**
In this commit, we describe the mechanism of sync point in Hinted Handoff in the user documentation. We explain the motivation for it and how to use it, as well as list and describe all of the parameters involved in the process. Errors that may appear and experienced by the user are addressed in the article.

Fixes scylladb/scylladb#18500

### CODE REVIEW

File: docs/architecture/anti-entropy/hinted-handoff.rst [hunk 1]
```diff
@@ -43,4 +43,34 @@ Hinted handoff is enabled and managed by these settings in :code:`scylla.yaml`:
...
+ScyllaDB provides you with a means to wait until hints for given nodes have been replayed. It's realized by creating a so-called sync point that corresponds to the hints that have accumulated up to the time of its creation. It's always created for both regular hints and updates for materialized views. You can create and await sync points by sending HTTP requests to :code:`http://{API_ADDRESS}:{API_PORT}/hinted_handoff/sync_point/`, where :code:`{API_ADDRESS}` corresponds to the API IP address of the coordinator node which hints are supposed to be sent from.
+
+You can create a sync point by sending a POST request to :code:`/hinted_handoff/sync_point` with the following parameters:
+
+* :code:`target_hosts`: a list of the IP addresses of nodes towards which hints should be replayed; the IPs are divided by commas, e.g. :code:`127.0.0.2,127.0.0.3`. If no IP address is specified, ScyllaDB assumes the sync point should be created towards all live nodes in the cluster. This parameter is not required -- if it's not present, ScyllaDB will treat the request as if zero IP addresses were specified.
+
+ScyllaDB should return a JSON containing the ID of the created sync point as the response.
```
Comment (piodul) [inline docs/architecture/anti-entropy/hinted-handoff.rst:55]: It would be nice if you provided an example here. You can use `curl` and copy-paste the relevant part from your terminal (i.e. the command and the output). This would make it easier to understand and would clarify some details that might have been omitted there, e.g. how the schema of the result JSON looks like.

File: docs/architecture/anti-entropy/hinted-handoff.rst [hunk 2]
```diff
@@ -43,4 +43,34 @@ Hinted handoff is enabled and managed by these settings in :code:`scylla.yaml`:
...
+
+* :code:`id`: the ID of the sync point the user wants to await. This parameter is mandatory. Only one ID can be passed,
+* :code:`timeout`: an integer value specifying how many seconds ScyllaDB should wait for the hints corresponding to the sync point to be sent. If the value is negative, ScyllaDB will wait indefinitely long. This parameter is not mandatory; if it's not present, ScyllaDB will return the result immediately.
+
+The database should return a JSON containing one of the two values as the response:
+
+* :code:`"DONE"` -- if all of the hints corresponding to the passed sync point have already been sent,
+* :code:`"IN_PROGRESS"` -- if some of those hints have not been sent yet.
+
+In case of an invalid request or an error when attempting to create or await a sync point, a JSON reporting the issue will be returned as the response.
```
Comment (piodul) [inline docs/architecture/anti-entropy/hinted-handoff.rst:67]: Again, some examples would be appreciated here.

**Review by piodul — CHANGES_REQUESTED** [review]

File: docs/architecture/anti-entropy/hinted-handoff.rst [hunk 3]
```diff
@@ -43,4 +43,34 @@ Hinted handoff is enabled and managed by these settings in :code:`scylla.yaml`:
...
+
+.. note:: If the node is turned off while you're awaiting a sync point on it, you'll receive an error JSON message as the response. You should try awaiting the sync point again when the node has been turned on again.
+
+Sync points aren't a fully reliable mechanism. If you receive :code:`"IN_PROGRESS"` as the response, you have a guarantee that not all of the hints corresponding to the passed sync point have been sent yet. However, if you receive :code:`"DONE"`, it may happen that **not** all of the nodes specified by :code:`target_hosts` when creating the sync point have received all of the hints.
```
[the remark this answers is not in the document]
Comment (dawmd) [reply docs/architecture/anti-entropy/hinted-handoff.rst:71]: Yes. Consider the following scenario:

1. Start node A with hinted handoff enabled for all DCs.
2. Save hints on node A for nodes B and C, where B belongs to the same DC as A – DC1, but C belongs to another one – DC2.
3. Create a sync point on node A for nodes B and C.
4. Turn off node A.
5. Start node A with hinted handoff enabled only for DC1. Because of that, only an endpoint manager corresponding to node B will be created.
6. Nodes B and C become reachable from node A again. Since there's an endpoint manager only for node B, only hints targeting that node will be sent.
7. The user awaits the previously created sync point and Scylla reports they've all been sent because for all existing endpoint managers, we've sent all of the hints saved until the moment when the sync point was created.

The issue is that we don't verify that there's an endpoint manager for every IP a sync point corresponds to, so we might end up checking only a subset of those.

I didn't give this much thought, but I think there might be a solution.

**Solution 1.** At the beginning of `manager::wait_for_sync_point()`, verify that we've created endpoint managers for all of the endpoints that the given sync point is concerned about. If not all of them exist, scan the disk and create the missing ones.

Potential problems:

1. We'd need to analyse if this would interact well during the transition period from IP-based to host-ID-based hinted handoff. Chances are this solutions might be difficult to implement in those circumstances because of the mappings IP–host ID we keep track of. For example, what if there's no mapping IP–host ID in the token metadata for a given IP? Should we ignore such a hint directory? Or should we not?

2. It could break the invariant that we only send hints to the DCs specified by the currently used host filter. This shouldn't be that big of an issue, i.e. I guess it would be fairly easy to implement, but it's definitely something to keep in mind.

**Solution 2.** Always create endpoint managers for all existing hint directories.

Potential problems: pretty much the same as above, except this strategy would proboably be easier to implement.

I think it would be best to leave the current implementation as-is until we get to 6.2 and can drop support for IP-based hinted handoff. We'd invalidate all of the created sync points before that version and change the parameter `target_hosts` to host IDs. Then solving the problems should be at least a bit easier.

When thinking about changing the implementation, one other scenario worth keeping in mind is this:

1. Start node A.
2. Save hints on node A for nodes B and C.
3. Create a sync point on node A for B and C.
4. B is, for some reason, drained on A and all of its hints are sent. The hint directory is deleted as well.
5. Restart A. An endpoint manager is created for C, but not for B (because there is no hint directory corresponding to it anymore).
6. C becomes reachable from A again and we send all of the hints to it.
7. The user awaits the sync point.

Currently, we'd (correctly) return `"DONE"` here.

[the remark this answers is not in the document]
Comment (piodul) [reply docs/architecture/anti-entropy/hinted-handoff.rst:71]: Please create an issue and post this description there. This PR's review is not the right place to discuss the details of the bug. The most important thing is that it's not a design limitation, it looks like this bug can be fixed.

Every feature has, or can have, bugs. We might encounter a bug where `IN_PROGRESS` keeps being returned even after all hints were sent. Are you willing to update the documentation every time new bugs appear?

(Btw, this property might not be considered true right now - imagine that it takes 100ms to send this value back to the client, but 50ms after the response was generated hints were replayed - I don't consider that a bug)

[the remark this answers is not in the document]
Comment (dawmd) [reply docs/architecture/anti-entropy/hinted-handoff.rst:71]: I see, that makes sense. I'll remove this fragment then.

**Comment by dawmd** [comment]
Version 2:
* Removed the fragment addressing potential bugs in the implementation.
* Added examples

File: docs/architecture/anti-entropy/hinted-handoff.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (dawmd) [reply docs/architecture/anti-entropy/hinted-handoff.rst]: Issue created: #19703

**Comment by piodul** [comment]
The iwyu job seems fixed in master, I'll rerun it.

File: docs/architecture/anti-entropy/hinted-handoff.rst [hunk 4]
```diff
@@ -43,4 +43,65 @@ Hinted handoff is enabled and managed by these settings in :code:`scylla.yaml`:
 
 Storing of the hint can also fail. Enabling hinted handoff therefore does not eliminate the need for repair; a user must recurrently :doc:`run a full repair </operating-scylla/procedures/maintenance/repair/>` to ensure data consistency across the cluster nodes.
 
+Sync point HTTP API
+^^^^^^^^^^^^^^^^^^^
+
+ScyllaDB provides you with a means to wait until hints for given nodes have been replayed. It's realized by creating a so-called sync point that corresponds to the hints that have accumulated up to the time of its creation. It's always created for both regular hints and updates for materialized views. You can create and await sync points by sending HTTP requests to :code:`http://{API_ADDRESS}:{API_PORT}/hinted_handoff/sync_point/`, where :code:`{API_ADDRESS}` corresponds to the API IP address of the coordinator node which hints are supposed to be sent from.
```
Comment (piodul) [inline docs/architecture/anti-entropy/hinted-handoff.rst:49]: > It's always created for both regular hints and updates for materialized views

The meaning of "created for" is not clear to me. I'd change it to "a sync point includes both ..." or something similar.

Moreover, contractions ("it's") shouldn't be used in formal language, please expand them.

File: docs/architecture/anti-entropy/hinted-handoff.rst [hunk 5]
```diff
@@ -43,4 +43,65 @@ Hinted handoff is enabled and managed by these settings in :code:`scylla.yaml`:
...
+
+Examples:
+
+.. code-block::
+
+    $ curl -X POST http://127.0.0.1:10000/hinted_handoff/sync_point?target_hosts=127.0.0.2,127.0.0.3
+    "Ay4AAADgR8IYRSM14vPuN232DLKSAQAMAAAAAAAAAAAAAAAMAAAAAAAAAAAAAAA/CncIC5VDTA=="
+
+    $ curl -X POST "http://127.0.0.1:10000/hinted_handoff/sync_point"
+    "Ay4AAADgR8IYRSM14vPuN232DLKSAQAMAAAAAAAAAAAAAAAMAAAAAAAAAAAAAAA/CncIC5VDTA=="
```
Comment (piodul) [inline docs/architecture/anti-entropy/hinted-handoff.rst:65]: Please add a short comment near each one of them to explain what they do and what kind of a sync point they create. I know that you described those cases in one of the paragraphs before this code block, but it's much easier to correlate a description with the example if it's right next to it.

File: docs/architecture/anti-entropy/hinted-handoff.rst [hunk 6]
```diff
@@ -43,4 +43,65 @@ Hinted handoff is enabled and managed by these settings in :code:`scylla.yaml`:
...
+    "DONE"
+
+    $ curl -X GET "http://127.0.0.1:10000/hinted_handoff/sync_point?timeout=-1"
+    {"message": "Missing mandatory parameter 'id'", "code": 400}
+
+    $ curl -X GET "http://127.0.0.1:10000/hinted_handoff/sync_point?id=invalid_id&timeout=-1"
+    {"message": "Failed to parse the sync point description invalid_id: Base64 encoded length is expected a multiple of 4 bytes but found: 10", "code": 400}
+
+    $ curl -X GET "http://127.0.0.1:10000/hinted_handoff/sync_point?id=invalidIdddd&timeout=-1"
+    {"message": "Failed to parse the sync point description invalidIdddd: Unsupported sync point format version: 138", "code": 400}
```
Comment (piodul) [inline docs/architecture/anti-entropy/hinted-handoff.rst:99]: Not sure if we should document these. It's rather obvious that you will get an error if you provide a malformed sync point ID. The error example in the previous block might be useful to explain how a potential error will look like, but the ones here are unnecessary in my opinion. No need for two of them, that's for sure.

File: docs/architecture/anti-entropy/hinted-handoff.rst [hunk 7]
```diff
@@ -43,4 +43,65 @@ Hinted handoff is enabled and managed by these settings in :code:`scylla.yaml`:
...
+
+    $ curl -X POST "http://127.0.0.1:10000/hinted_handoff/sync_point?target_hosts=invalid_host"
+    {"message": "Failed to parse host address invalid_host: invalid_host", "code": 400}
+
+You can await a previously created sync point by sending a GET request to :code:`/hinted_handoff/sync_point` with the following parameters:
+
+* :code:`id`: the ID of the sync point the user wants to await. This parameter is mandatory. Only one ID can be passed,
+* :code:`timeout`: an integer value specifying how many seconds ScyllaDB should wait for the hints corresponding to the sync point to be sent. If the value is negative, ScyllaDB will wait indefinitely long. This parameter is not mandatory; if it's not present, ScyllaDB will return the result immediately.
+
+The database should return a JSON containing one of the two values as the response:
```
Comment (piodul) [inline docs/architecture/anti-entropy/hinted-handoff.rst:75]: Nit: let's be more specific: a JSON string.

File: docs/architecture/anti-entropy/hinted-handoff.rst [hunk 8]
```diff
@@ -43,4 +43,65 @@ Hinted handoff is enabled and managed by these settings in :code:`scylla.yaml`:
...
+
+You can create a sync point by sending a POST request to :code:`/hinted_handoff/sync_point` with the following parameters:
+
+* :code:`target_hosts`: a list of the IP addresses of nodes towards which hints should be replayed; the IPs are divided by commas, e.g. :code:`127.0.0.2,127.0.0.3`. If no IP address is specified, ScyllaDB assumes the sync point should be created towards all live nodes in the cluster. This parameter is not required -- if it's not present, ScyllaDB will treat the request as if zero IP addresses were specified.
+
+ScyllaDB should return a JSON containing the ID of the created sync point as the response. In case of an error, a JSON reporting the issue will be returned instead.
```
Comment (piodul) [inline docs/architecture/anti-entropy/hinted-handoff.rst:55]: > ScyllaDB should return a JSON containing ... In case of an error, a JSON reporting ...

Nit: a "JSON object", not just "JSON". JSON is a language / data format.

File: docs/architecture/anti-entropy/hinted-handoff.rst [hunk 9]
```diff
@@ -43,4 +43,65 @@ Hinted handoff is enabled and managed by these settings in :code:`scylla.yaml`:
...
 
+Sync point HTTP API
+^^^^^^^^^^^^^^^^^^^
+
+ScyllaDB provides you with a means to wait until hints for given nodes have been replayed. It's realized by creating a so-called sync point that corresponds to the hints that have accumulated up to the time of its creation. It's always created for both regular hints and updates for materialized views. You can create and await sync points by sending HTTP requests to :code:`http://{API_ADDRESS}:{API_PORT}/hinted_handoff/sync_point/`, where :code:`{API_ADDRESS}` corresponds to the API IP address of the coordinator node which hints are supposed to be sent from.
+
+You can create a sync point by sending a POST request to :code:`/hinted_handoff/sync_point` with the following parameters:
+
+* :code:`target_hosts`: a list of the IP addresses of nodes towards which hints should be replayed; the IPs are divided by commas, e.g. :code:`127.0.0.2,127.0.0.3`. If no IP address is specified, ScyllaDB assumes the sync point should be created towards all live nodes in the cluster. This parameter is not required -- if it's not present, ScyllaDB will treat the request as if zero IP addresses were specified.
```
Comment (piodul) [inline docs/architecture/anti-entropy/hinted-handoff.rst:53]: > :code:`target_hosts`: a list of the IP addresses of nodes towards which hints should be replayed

It's a bit vague in my opinion, this could suggest that creating a sync point waits for something. I'd rephrase it, say something about "including" hints targeted to those specific hosts. To make your point clear, you can also say what will awaiting a sync point created for a given `target_hosts` will do.

File: docs/architecture/anti-entropy/hinted-handoff.rst [hunk 10]
```diff
@@ -43,4 +43,65 @@ Hinted handoff is enabled and managed by these settings in :code:`scylla.yaml`:
...
+    {"message": "Missing mandatory parameter 'id'", "code": 400}
+
+    $ curl -X GET "http://127.0.0.1:10000/hinted_handoff/sync_point?id=invalid_id&timeout=-1"
+    {"message": "Failed to parse the sync point description invalid_id: Base64 encoded length is expected a multiple of 4 bytes but found: 10", "code": 400}
+
+    $ curl -X GET "http://127.0.0.1:10000/hinted_handoff/sync_point?id=invalidIdddd&timeout=-1"
+    {"message": "Failed to parse the sync point description invalidIdddd: Unsupported sync point format version: 138", "code": 400}
+
+
+.. note:: If the node is turned off while you're awaiting a sync point on it, you'll receive an error JSON message as the response. You should try awaiting the sync point again when the node has been turned on again.
```
Comment (piodul) [inline docs/architecture/anti-entropy/hinted-handoff.rst:102]: > you'll receive an error JSON message as the response

Not sure if it's necessary to be so specific that you will receive an "error JSON message". You can just say that you will get an error if this happens. JSON is just a format for encoding the error, it's already explained in previous sections.

File: docs/architecture/anti-entropy/hinted-handoff.rst [hunk 11]
```diff
@@ -43,4 +43,65 @@ Hinted handoff is enabled and managed by these settings in :code:`scylla.yaml`:
...
+
+.. note:: When the whole ScyllaDB cluster has finished upgrading from version **5.4** to version **6.0**, Hinted Handoff will undergo a transition from IP-based to host-ID-based. During that process, Hinted Handoff will be disabled for a short time. As a consequence, if you're awaiting a sync point at that time, you might receive an error JSON message. Try to await the sync point again when the transition has finished.
```
Comment (piodul) [inline docs/architecture/anti-entropy/hinted-handoff.rst:104]: > an error JSON message

Ditto

File: docs/architecture/anti-entropy/hinted-handoff.rst [hunk 12]
```diff
@@ -43,4 +43,65 @@ Hinted handoff is enabled and managed by these settings in :code:`scylla.yaml`:
...
+    {"message": "Failed to parse host address invalid_host: invalid_host", "code": 400}
+
+You can await a previously created sync point by sending a GET request to :code:`/hinted_handoff/sync_point` with the following parameters:
+
+* :code:`id`: the ID of the sync point the user wants to await. This parameter is mandatory. Only one ID can be passed,
+* :code:`timeout`: an integer value specifying how many seconds ScyllaDB should wait for the hints corresponding to the sync point to be sent. If the value is negative, ScyllaDB will wait indefinitely long. This parameter is not mandatory; if it's not present, ScyllaDB will return the result immediately.
+
+The database should return a JSON containing one of the two values as the response:
+
+* :code:`"DONE"` -- if all of the hints corresponding to the passed sync point have already been sent,
```
Comment (piodul) [inline docs/architecture/anti-entropy/hinted-handoff.rst:77]: Nit: passing can also mean "go past something", and you could argue that "passing a sync point" could be interpreted in this second, alternative way. Let's avoid possible confusion and change "passed" to a word which doesn't have this second meaning, e.g. "provided".

File: docs/architecture/anti-entropy/hinted-handoff.rst [hunk 13]
```diff
@@ -43,4 +43,65 @@ Hinted handoff is enabled and managed by these settings in :code:`scylla.yaml`:
...
+* :code:`"IN_PROGRESS"` -- if some of those hints have not been sent yet.
+
+In case of an error, a JSON reporting the issue will be returned instead.
```
Comment (piodul) [inline docs/architecture/anti-entropy/hinted-handoff.rst:80]: > a JSON reporting the issue

Ditto about the "JSON".

**Review by piodul — CHANGES_REQUESTED** [review]

**Comment by dawmd** [comment]
> The iwyu job seems fixed in master, I'll rerun it.

Unfortunately, it needs rebasing. I'll do it

**Comment by dawmd** [comment]
Version 2:
* Rebased the PR on top of the current master branch (to prevent CI failures).
* Avoided contractions in the documentation.
* Clarified some fragments that might've been ambiguous.
* Enabled syntax highlighting in the code snippets.
* Added comments describing the commends in the code snippets.

**Review by piodul — APPROVED** [review]
LGTM

@annastuchlik could you review it as well?

**Comment by piodul** [comment]
> LGTM
> 
> @annastuchlik could you review it as well?

@annastuchlik review ping

**Review by nyh — APPROVED** [review]

**Comment by nyh** [comment]
@piodul I see you pinged @annastuchlik 15 minutes ago for her review, so I won't merge yet.

I think it's very good that we add this documentation, but I wonder if it's a feature that end-users are really supposed to use or even know about, or if we only plan it to be used in test and automation (e.g., Scylla Cloud). If it's the latter, maybe it should be in the dev/ subdirectory?

**Comment by piodul** [comment]
> @piodul I see you pinged @annastuchlik 15 minutes ago for her review, so I won't merge yet.
> 
> I think it's very good that we add this documentation, but I wonder if it's a feature that end-users are really supposed to use or even know about, or if we only plan it to be used in test and automation (e.g., Scylla Cloud). If it's the latter, maybe it should be in the dev/ subdirectory?

Hmm, that's a good question - I'm not sure. It's a niche feature, but I suppose it can be useful in some cases for users. Originally, it was developed in order to let hints get replayed before repair runs because it can sometimes make the latter repair cheaper. I'm not sure whether any user ever asked for this, but I do know that field engineers asked about this feature some time ago (not knowing that it was already implemented).

File: docs/architecture/anti-entropy/hinted-handoff.rst [hunk 14]
```diff
@@ -43,4 +43,65 @@ Hinted handoff is enabled and managed by these settings in :code:`scylla.yaml`:
...
+* :code:`target_hosts`: a list of the IP addresses of nodes to be covered by the sync point; the IPs are divided by commas, e.g. :code:`127.0.0.2,127.0.0.3`. If no IP address is specified, ScyllaDB assumes the sync point should be created towards all nodes in the cluster. This parameter is not required -- if it is not present, ScyllaDB will treat the request as if zero IP addresses were specified.
+
+ScyllaDB should return a JSON string containing the ID of the created sync point as the response. In case of an error, a JSON object reporting the issue will be returned instead.
+
+Examples:
+
+.. code-block:: bash
+
+    # Create a sync point for hosts of IP addresses: 127.0.0.2 and 127.0.0.3.
+    $ curl -X POST "http://127.0.0.1:10000/hinted_handoff/sync_point?target_hosts=127.0.0.2,127.0.0.3"
```
Comment (bhalevy) [inline docs/architecture/anti-entropy/hinted-handoff.rst:62]: Do we accept also host_id:s?
I hope so, now that hinted_handoff is supposed to transition to using host_id:s rather than endpoint addresses.

Comment (bhalevy) [reply docs/architecture/anti-entropy/hinted-handoff.rst:62]: https://github.com/scylladb/scylladb/issues/19924

Comment (dawmd) [reply docs/architecture/anti-entropy/hinted-handoff.rst:62]: > Do we accept also host_id:s?
We don't, but we'll have to do it

**Comment by dawmd** [comment]
Version 3:
* Dropped the note about migrating from IP-based to host-ID-based hinted handoff and its consequences on sync points.
* Updated the commit and PR messages.

**Comment by piodul** [comment]
Queued

File: docs/architecture/anti-entropy/hinted-handoff.rst [hunk 15]
```diff
@@ -43,4 +43,63 @@ Hinted handoff is enabled and managed by these settings in :code:`scylla.yaml`:
 
 Storing of the hint can also fail. Enabling hinted handoff therefore does not eliminate the need for repair; a user must recurrently :doc:`run a full repair </operating-scylla/procedures/maintenance/repair/>` to ensure data consistency across the cluster nodes.
 
+Sync point HTTP API
+^^^^^^^^^^^^^^^^^^^
```
Comment (tzach) [inline docs/architecture/anti-entropy/hinted-handoff.rst:47]: We have API reference we can point to https://opensource.docs.scylladb.com/stable/reference/api-reference.html#hinted-handoff

(can be in a followup PR)

Comment (dawmd) [reply docs/architecture/anti-entropy/hinted-handoff.rst:47]: Just to make sure I understand – you want to have the header be a link to that article, correct?

Comment (tzach) [reply docs/architecture/anti-entropy/hinted-handoff.rst:47]: > Just to make sure I understand – you want to have the header be a link to that article, correct? 

Yes, it would be useful to let users know about the API reference.

Comment (dawmd) [reply docs/architecture/anti-entropy/hinted-handoff.rst:47]: scylladb/scylladb#20065
